File: backend-central/routes/camera_routes.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import Optional
import yaml
import os

router = APIRouter(prefix="/api/rtsp-cameras", tags=["rtsp-cameras"])

# go2rtc config file path
CONFIG_FILE = os.path.join(os.path.dirname(os.path.dirname(__file__)), "go2rtc.yaml")

class Camera(BaseModel):
    id: str
    name: str
    type: str
    url: str
    hasAudio: bool = False

class CameraUpdate(BaseModel):
    url: Optional[str] = None
    name: Optional[str] = None
    type: Optional[str] = None
    newId: Optional[str] = None

@router.get("/")
async def get_cameras():
    """Get all cameras from go2rtc.yaml"""
    if not os.path.exists(CONFIG_FILE):
        return []

    try:
        with open(CONFIG_FILE, 'r', encoding='utf-8') as f:
            config = yaml.safe_load(f) or {}
    except Exception as e:
        print(f"Error reading go2rtc.yaml: {e}")
        return []

    streams = config.get('streams', {})
    metadata = config.get('metadata', {})

    cameras = []
    for cam_id, url_value in streams.items():
        if cam_id.startswith('#'):
            continue

        # Remove go2rtc params
        url = url_value.split('#')[0] if '#' in str(url_value) else str(url_value)

        meta = metadata.get(cam_id, {})
        cameras.append({
            'id': cam_id,
            'name': meta.get('name', cam_id.replace('_', ' ').title()),
            'type': meta.get('type', 'rtsp' if url.startswith('rtsp://') else 'public'),
            'url': url,
            'hasAudio': meta.get('hasAudio', False)
        })

    return cameras


@router.post("/")
async def add_camera(camera: Camera):
    """Add new camera to go2rtc.yaml"""
    # Read config
    config = {}
    if os.path.exists(CONFIG_FILE):
        try:
            with open(CONFIG_FILE, 'r', encoding='utf-8') as f:
                config = yaml.safe_load(f) or {}
        except Exception as e:
            raise HTTPException(status_code=500, detail=f"Error reading config: {e}")

    if 'streams' not in config:
        config['streams'] = {}
    if 'metadata' not in config:
        config['metadata'] = {}

    # Check duplicate
    if camera.id in config['streams']:
        raise HTTPException(
            status_code=400,
            detail=f"Camera with name '{camera.name}' already exists. Please use a different name."
        )

    # Add camera
    config['streams'][camera.id] = camera.url
    config['metadata'][camera.id] = {
        'name': camera.name,
        'type': camera.type,
        'hasAudio': camera.hasAudio
    }

    # Write config
    try:
        os.makedirs(os.path.dirname(CONFIG_FILE), exist_ok=True)
        with open(CONFIG_FILE, 'w', encoding='utf-8') as f:
            yaml.dump(config, f, default_flow_style=False, allow_unicode=True)
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error writing config: {e}")

    return {"success": True, "message": "Camera added successfully"}


@router.put("/{cam_id}")
async def update_camera(cam_id: str, update: CameraUpdate):
    """Update camera in go2rtc.yaml"""
    if not os.path.exists(CONFIG_FILE):
        raise HTTPException(status_code=404, detail="Config file not found")

    try:
        with open(CONFIG_FILE, 'r', encoding='utf-8') as f:
            config = yaml.safe_load(f) or {}
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error reading config: {e}")

    if 'streams' not in config or cam_id not in config['streams']:
        raise HTTPException(status_code=404, detail="Camera not found")

    # Handle ID change
    target_id = cam_id
    if update.newId and update.newId != cam_id:
        if update.newId in config['streams']:
            raise HTTPException(
                status_code=400,
                detail=f"Camera with name '{update.name or update.newId}' already exists"
            )

        # Migrate to new ID
        config['streams'][update.newId] = config['streams'][cam_id]
        config['metadata'][update.newId] = config.get('metadata', {}).get(cam_id, {})
        del config['streams'][cam_id]
        if cam_id in config.get('metadata', {}):
            del config['metadata'][cam_id]
        target_id = update.newId

    # Update URL
    if update.url:
        config['streams'][target_id] = update.url

    # Update metadata
    if 'metadata' not in config:
        config['metadata'] = {}
    if target_id not in config['metadata']:
        config['metadata'][target_id] = {}

    if update.name is not None:
        config['metadata'][target_id]['name'] = update.name
    if update.type is not None:
        config['metadata'][target_id]['type'] = update.type

    # Write config
    try:
        with open(CONFIG_FILE, 'w', encoding='utf-8') as f:
            yaml.dump(config, f, default_flow_style=False, allow_unicode=True)
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error writing config: {e}")

    return {
        "success": True,
        "message": "Camera updated successfully",
        "id": target_id
    }


@router.delete("/{cam_id}")
async def delete_camera(cam_id: str):
    """Remove camera from go2rtc.yaml"""
    if not os.path.exists(CONFIG_FILE):
        raise HTTPException(status_code=404, detail="Config file not found")

    try:
        with open(CONFIG_FILE, 'r', encoding='utf-8') as f:
            config = yaml.safe_load(f) or {}
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error reading config: {e}")

    if 'streams' not in config or cam_id not in config['streams']:
        raise HTTPException(status_code=404, detail="Camera not found")

    # Remove camera
    del config['streams'][cam_id]
    if 'metadata' in config and cam_id in config['metadata']:
        del config['metadata'][cam_id]

    # Write config
    try:
        with open(CONFIG_FILE, 'w', encoding='utf-8') as f:
            yaml.dump(config, f, default_flow_style=False, allow_unicode=True)
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error writing config: {e}")

    return {"success": True, "message": "Camera removed successfully"}

File: backend/routes/camera.routes.js
const express = require("express");
const {
  readConfig,
  writeConfig,
  addStreamToRuntime,
  removeStreamFromRuntime,
} = require("../helpers/config");

const router = express.Router();

// GET /api/cameras - Get all cameras
router.get("/", (req, res) => {
  try {
    const config = readConfig();
    const streams = config.streams || {};
    const metadata = config.metadata || {};

    const cameras = Object.entries(streams)
      .filter(([key]) => !key.startsWith("#")) // Ignore comments
      .map(([id, urlValue]) => {
        // Remove go2rtc params from URL when returning to frontend
        let url = urlValue;
        if (url && url.includes("#")) {
          url = url.split("#")[0];
        }
        const meta = metadata[id] || {};
        const hasAudio = meta.hasAudio === true;

        return {
          id,
          name:
            meta.name ||
            id.replace(/_/g, " ").replace(/\b\w/g, (l) => l.toUpperCase()),
          type:
            meta.type || (url && url.startsWith("rtsp://") ? "rtsp" : "public"),
          url,
          hasAudio,
        };
      });

    res.json(cameras);
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// POST /api/cameras - Add a new camera
router.post("/", async (req, res) => {
  try {
    const { id, url, name, type } = req.body;

    if (!id || !url) {
      return res.status(400).json({ error: "Missing id or url" });
    }

    const config = readConfig();

    // Initialize streams and metadata if not exists
    if (!config.streams) {
      config.streams = {};
    }
    if (!config.metadata) {
      config.metadata = {};
    }

    // Check if camera with this id already exists
    if (config.streams[id]) {
      return res.status(400).json({
        error: `Camera with name "${
          name || id
        }" already exists. Please use a different name.`,
      });
    }

    // Store URL as-is
    let finalUrl = url;

    // Add new camera stream
    config.streams[id] = finalUrl;

    // Add camera metadata
    config.metadata[id] = {
      name: name || id,
      type: type || "rtsp",
      hasAudio: false, // Default: no audio
    };

    // Write to file
    writeConfig(config);

    res.json({ success: true, message: "Camera added successfully" });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// PUT /api/cameras/:id - Update a camera
router.put("/:id", async (req, res) => {
  try {
    const { id } = req.params;
    const { url, name, type, newId } = req.body;

    const config = readConfig();

    if (!config.streams || !config.streams[id]) {
      return res.status(404).json({ error: "Camera not found" });
    }

    // If name changed, we need to update the ID
    let targetId = id;
    if (newId && newId !== id) {
      // Check if new ID already exists
      if (config.streams[newId]) {
        return res.status(400).json({
          error: `Camera with name "${
            name || newId
          }" already exists. Please use a different name.`,
        });
      }

      // Migrate camera to new ID
      config.streams[newId] = config.streams[id];
      config.metadata[newId] = config.metadata[id] || {};
      delete config.streams[id];
      delete config.metadata[id];
      targetId = newId;

      // Migrate stream in go2rtc runtime
      try {
        await removeStreamFromRuntime(id);
        await addStreamToRuntime(newId, config.streams[newId]);
      } catch (error) {
        console.error("Error migrating stream in go2rtc:", error);
      }
    }

    // Update camera stream if URL changed
    if (url) {
      let finalUrl = url;
      config.streams[targetId] = finalUrl;

      // Update stream in go2rtc runtime
      try {
        await addStreamToRuntime(targetId, finalUrl);
      } catch (error) {
        console.error("Error updating stream in go2rtc:", error);
      }
    }

    // Update camera metadata
    if (!config.metadata) {
      config.metadata = {};
    }
    if (!config.metadata[targetId]) {
      config.metadata[targetId] = {};
    }

    if (name !== undefined) {
      config.metadata[targetId].name = name;
      if (config.metadata[targetId].hasAudio === undefined) {
        config.metadata[targetId].hasAudio = false;
      }
    }
    if (type !== undefined) {
      config.metadata[targetId].type = type;
    }

    // Write to file
    writeConfig(config);

    res.json({
      success: true,
      message: "Camera updated successfully",
      id: targetId,
    });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// DELETE /api/cameras/:id - Remove a camera
router.delete("/:id", async (req, res) => {
  try {
    const { id } = req.params;

    const config = readConfig();

    if (!config.streams || !config.streams[id]) {
      return res.status(404).json({ error: "Camera not found" });
    }

    // Remove camera stream
    delete config.streams[id];

    // Remove camera metadata
    if (config.metadata && config.metadata[id]) {
      delete config.metadata[id];
    }

    // Write to file
    writeConfig(config);

    res.json({ success: true, message: "Camera removed successfully" });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

module.exports = router;
